src/consumption.cpp

```diff
@@ -101,69 +101,120 @@ int player::stomach_capacity() const
+    // As float to avoid rounding too many times
+    float kcal = comest.get_comestible()->default_nutrition.kcal;
 
-        // Many raw foods give less calories, as your body has expends more energy digesting them.
-        // We don't want RAW to stack for components and results, so we're doing it in this else block.
-        if( comest.has_flag( "RAW" ) && !comest.has_flag( "COOKED" ) ) {
-            kcal *= 0.75f;
-        }
+    // Many raw foods give less calories, as your body has expends more energy digesting them.
+    if( comest.has_flag( "RAW" ) && !comest.has_flag( "COOKED" ) ) {
+        kcal *= 0.75f;
     }
 
-    if( has_trait( trait_GIZZARD ) ) {
+    if( p.has_trait( trait_GIZZARD ) ) {
         kcal *= 0.6f;
     }
 
-    if( has_trait( trait_CARNIVORE ) && comest.has_flag( flag_CARNIVORE_OK ) &&
+    if( p.has_trait( trait_CARNIVORE ) && comest.has_flag( flag_CARNIVORE_OK ) &&
         comest.has_any_flag( carnivore_blacklist ) ) {
         // TODO: Comment pizza scrapping
         kcal *= 0.5f;
     }
 
     const float relative_rot = comest.get_relative_rot();
     // Saprophages get full nutrition from rotting food
-    if( relative_rot > 1.0f && !has_trait( trait_SAPROPHAGE ) ) {
+    if( relative_rot > 1.0f && !p.has_trait( trait_SAPROPHAGE ) ) {
         // everyone else only gets a portion of the nutrition
         // Scaling linearly from 100% at just-rotten to 0 at halfway-rotten-away
         const float rottedness = clamp( 2 * relative_rot - 2.0f, 0.1f, 1.0f );
         kcal *= ( 1.0f - rottedness );
     }
 
     // Bionic digestion gives extra nutrition
-    if( has_bionic( bio_digestion ) ) {
+    if( p.has_bionic( bio_digestion ) ) {
         kcal *= 1.5f;
     }
 
     return static_cast<int>( kcal );
 }
 
+// Compute default effective vitamins for an item, taking into account player
+// traits, but not components of the item.
+static std::map<vitamin_id, int> compute_default_effective_vitamins(
+    const item &it, const player &p )
+{
+    assert( it.get_comestible() );
+
+    std::map<vitamin_id, int> res = it.get_comestible()->default_nutrition.vitamins;
+
+    for( const trait_id &trait : p.get_mutations() ) {
+        const auto &mut = trait.obj();
+        // make sure to iterate over every material defined for vitamin absorption
+        // TODO: put this loop into a function and utilize it again for bionics
+        for( const auto &mat : mut.vitamin_absorb_multi ) {
+            // this is where we are able to check if the food actually is changed by the trait
+            if( mat.first == material_id( "all" ) || it.made_of( mat.first ) ) {
+                const std::map<vitamin_id, double> &mat_vit_map = mat.second;
+                for( const auto &vit : res ) {
+                    auto vit_factor = mat_vit_map.find( vit.first );
+                    if( vit_factor != mat_vit_map.end() ) {
+                        res[ vit.first ] *= vit_factor->second;
```

Comment:
Change the loop to iterator over non-const references (`for( auto &vit : `) and avoid the repeated look up here:
```suggestion
                        vit.second *= vit_factor->second;
```

Reply:
Done.